### ISSUE TITLE
confusing error message from front-end related to mispatched function types

### ISSUE THREAD

**Issue by mraleph** [dart-lang/sdk#34096]
```
void foo<T extends num>() {
  print(T);
}

void bar(void Function<T extends String>() f) {
  f();
}

main() {
  bar(foo);
}
```

CFE reports:

```
Error: The argument type '<T extends dart.core::num = dart.core::num>() → void' can't be assigned to the parameter type '<T extends dart.core::String = dynamic>() → void'.
Try changing the type of the parameter, or casting the argument to '<T extends dart.core::String = dynamic>() → void'
```

These `dart.core::num = dart.core::num` and `dart.core::String = dynamic` are extremely confusing because they don't map to any valid Dart syntax.

**Comment by kmillikin** [comment]
This encapsulates a lot of our problems (except the `#lib1` `#lib2` issues).

Ideally we'd get produce something like `void Function<T extends num>()` can't be assigned to `void Function<T extends String>()` because the bound `String` can't be assigned to the bound `num`.

We'll try to make incremental improvements when we can, and embark on a very concerted effort to improve errors after Dart 2.1.

**Comment by askeksa-google** [comment]
Fixed by c959bd98afb1d24cd944388040fa913a19021620.
